### PR TITLE
[Refactor] Refine create table

### PR DESCRIPTION
**Pull request by sduzh**
Use abstract factory design pattern to split the table creation into different classes based on the table engine;

## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/server/MysqlTableFactory.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+public class MysqlTableFactory implements AbstractTableFactory {
+
+    public static final MysqlTableFactory INSTANCE = new MysqlTableFactory();
+
+    private MysqlTableFactory() {
+    }
+
+    @Override
+    @NotNull
+    public Table createTable(LocalMetastore metastore, Database database, CreateTableStmt stmt) throws DdlException {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/server/MysqlTableFactory.java:36]: Moved from `LocalMetaStore.createMysqlTable` without any logical change.

File: fe/fe-core/src/main/java/com/starrocks/server/IcebergTableFactory.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+public class IcebergTableFactory extends ExternalTableFactory {
+    public static final IcebergTableFactory INSTANCE = new IcebergTableFactory();
+
+    private IcebergTableFactory() {
+
+    }
+
+    @Override
+    @NotNull
+    public Table createTable(LocalMetastore metastore, Database database, CreateTableStmt stmt) throws DdlException {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/server/IcebergTableFactory.java:45]: Moved from `TableFactory.createIcebergTable()` without any logical change.

File: fe/fe-core/src/main/java/com/starrocks/server/IcebergTableFactory.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            String partitionCmt = "PARTITION BY (" + String.join(", ", icebergTable.getPartitionColumnNames()) + ")";
+            icebergTable.setComment(partitionCmt);
+        } else if (!Strings.isNullOrEmpty(stmt.getComment())) {
+            icebergTable.setComment(stmt.getComment());
+        }
+
+        return icebergTable;
+    }
+
+    private static void validateIcebergColumnType(List<Column> columns, IcebergTable oTable) throws DdlException {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/server/IcebergTableFactory.java:87]: Moved from 'TableFactory.teIcebergColumnType()` without any change.

File: fe/fe-core/src/main/java/com/starrocks/server/ElasticSearchTableFactory.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    public static final ElasticSearchTableFactory INSTANCE = new ElasticSearchTableFactory();
+
+    private ElasticSearchTableFactory() {
+
+    }
+
+    @Override
+    @NotNull
+    public Table createTable(LocalMetastore metastore, Database database, CreateTableStmt stmt) throws DdlException {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/server/ElasticSearchTableFactory.java:42]: Moved from 'TableFactory.createESTable()` without any change.

File: fe/fe-core/src/main/java/com/starrocks/server/FileTableFactory.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+public class FileTableFactory implements AbstractTableFactory {
+    public static final FileTableFactory INSTANCE = new FileTableFactory();
+
+    private FileTableFactory() {
+
+    }
+
+    @Override
+    @NotNull
+    public Table createTable(LocalMetastore metastore, Database database, CreateTableStmt stmt) throws DdlException {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/server/FileTableFactory.java:38]: Moved from `TableFactory.createFileTable()` without any logical change.

File: fe/fe-core/src/main/java/com/starrocks/server/HiveTableFactory.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+public class HiveTableFactory extends ExternalTableFactory {
+    public static final HiveTableFactory INSTANCE = new HiveTableFactory();
+
+    private HiveTableFactory() {
+
+    }
+
+    @Override
+    @NotNull
+    public Table createTable(LocalMetastore metastore, Database database, CreateTableStmt stmt) throws DdlException {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/server/HiveTableFactory.java:44]: Moved from `TableFactory.createHiveTable()` without any logical change.

File: fe/fe-core/src/main/java/com/starrocks/server/HudiTableFactory.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+public class HudiTableFactory extends ExternalTableFactory {
+    public static final HudiTableFactory INSTANCE = new HudiTableFactory();
+
+    private HudiTableFactory() {
+
+    }
+
+    @Override
+    @NotNull
+    public Table createTable(LocalMetastore metastore, Database database, CreateTableStmt stmt) throws DdlException {
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/server/HudiTableFactory.java:52]: Moved from `TableFactory.createHudiTable()` without any logical change.

File: fe/fe-core/src/main/java/com/starrocks/server/IcebergTableFactory.java (hunk 2 above)
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/server/IcebergTableFactory.java:45]: Moved from `TableFactory.createIcebergTable()` without any logical change.

File: fe/fe-core/src/main/java/com/starrocks/server/AbstractTableFactory.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import com.starrocks.catalog.Database;
+import com.starrocks.catalog.Table;
+import com.starrocks.common.DdlException;
+import com.starrocks.sql.ast.CreateTableStmt;
+
+import javax.validation.constraints.NotNull;
+
+public interface AbstractTableFactory {
+    @NotNull
+    Table createTable(LocalMetastore metastore, Database database, CreateTableStmt stmt) throws DdlException;
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/server/AbstractTableFactory.java:26]: How about putting LocalMetastore into OlapTableFactory

File: fe/fe-core/src/main/java/com/starrocks/server/ExternalTableFactory.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.Map;
+
+import static com.starrocks.server.CatalogMgr.ResourceMappingCatalog.getResourceMappingCatalogName;
+
+public abstract class ExternalTableFactory implements AbstractTableFactory {
+    public static final String DB = "database";
+    public static final String TABLE = "table";
+    public static final String RESOURCE = "resource";
+    public static final String PROPERTY_MISSING_MSG =
+            "Hive %s is null. Please add properties('%s'='xxx') when create table";
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/server/ExternalTableFactory.java:32]: Hive -> Remote. thks

File: fe/fe-core/src/main/java/com/starrocks/server/AbstractTableFactory.java (hunk 8 above)
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/server/AbstractTableFactory.java:26]: Do you mean refactor the signature to `Table createTable(LocalMetastore metastore, Database database, CreateTableStmt stmt)`?

File: fe/fe-core/src/main/java/com/starrocks/server/ExternalTableFactory.java (hunk 9 above)
Comment (sduzh) [reply fe/fe-core/src/main/java/com/starrocks/server/ExternalTableFactory.java]: Fixed, PTAL

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 124 / 250 (49.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/IcebergTableFactory.java|0|44|00.00%|[37, 46, 47, 48, 49, 50, 52, 55, 56, 57, 58, 61, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 74, 77, 78, 79, 80, 81, 84, 88, 89, 90, 91, 92, 95, 96, 97, 100, 102, 103, 106, 107, 109, 110]|
|:large_blue_circle:|com/starrocks/server/ElasticSearchTableFactory.java|0|16|00.00%|[34, 43, 46, 47, 50, 51, 52, 53, 54, 56, 58, 59, 62, 63, 64, 65]|
|:large_blue_circle:|com/starrocks/server/HudiTableFactory.java|16|65|24.62%|[72, 73, 74, 77, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 92, 95, 96, 97, 98, 99, 102, 109, 110, 113, 114, 115, 116, 118, 119, 120, 121, 123, 124, 125, 127, 129, 130, 133, 134, 135, 137, 139, 140, 141, 142, 144, 145]|
|:large_blue_circle:|com/starrocks/server/TableFactoryProvider.java|13|18|72.22%|[21, 37, 45, 46, 48]|
|:large_blue_circle:|com/starrocks/server/HiveTableFactory.java|39|47|82.98%|[52, 53, 54, 80, 93, 98, 99, 104]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|6|7|85.71%|[2457]|
|:large_blue_circle:|com/starrocks/server/ExternalTableFactory.java|21|23|91.30%|[40, 74]|
|:large_blue_circle:|com/starrocks/server/FileTableFactory.java|14|15|93.33%|[54]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/MysqlTableFactory.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/JDBCTableFactory.java|6|6|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/server/AbstractTableFactory.java (hunk 8 above)
Comment (stephen-shelby) [reply fe/fe-core/src/main/java/com/starrocks/server/AbstractTableFactory.java:26]: yes, but manually creating the external table is currently only for compatibility.  we will recommend using the catalog to query external table. So I think if the this part is mainly for inner tables in future, passing it to localMetastore is fine.

**Review by stephen-shelby — DISMISSED** [review]

**Review by wyb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 124 / 250 (49.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/server/IcebergTableFactory.java|0|44|00.00%|[37, 46, 47, 48, 49, 50, 52, 55, 56, 57, 58, 61, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 74, 77, 78, 79, 80, 81, 84, 88, 89, 90, 91, 92, 95, 96, 97, 100, 102, 103, 106, 107, 109, 110]|
|:large_blue_circle:|com/starrocks/server/ElasticSearchTableFactory.java|0|16|00.00%|[34, 43, 46, 47, 50, 51, 52, 53, 54, 56, 58, 59, 62, 63, 64, 65]|
|:large_blue_circle:|com/starrocks/server/HudiTableFactory.java|16|65|24.62%|[72, 73, 74, 77, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 92, 95, 96, 97, 98, 99, 102, 109, 110, 113, 114, 115, 116, 118, 119, 120, 121, 123, 124, 125, 127, 129, 130, 133, 134, 135, 137, 139, 140, 141, 142, 144, 145]|
|:large_blue_circle:|com/starrocks/server/TableFactoryProvider.java|13|18|72.22%|[21, 37, 45, 46, 48]|
|:large_blue_circle:|com/starrocks/server/HiveTableFactory.java|39|47|82.98%|[52, 53, 54, 80, 93, 98, 99, 104]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|6|7|85.71%|[2457]|
|:large_blue_circle:|com/starrocks/server/ExternalTableFactory.java|21|23|91.30%|[40, 74]|
|:large_blue_circle:|com/starrocks/server/FileTableFactory.java|14|15|93.33%|[54]|
|:large_blue_circle:|com/starrocks/server/OlapTableFactory.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/MysqlTableFactory.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/JDBCTableFactory.java|6|6|100.00%|[]|

**Comment by sduzh** [comment]
@wyb @stephen-shelby I forgot to sign off commit and the DCO check failed, now it's fixed and force-pushed, please approve this PR again, sorry for that.

**Review by stephen-shelby — APPROVED** [review]

**Comment by sduzh** [comment]
run starrocks_admit_test

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**